src/algorithms/calorimetry/SimCalorimeterHitProcessor.cc

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2025 Minho Kim, Sylvester Joosten, Derek Anderson
+
+#include "SimCalorimeterHitProcessor.h"
+
+#include <DD4hep/Detector.h>
+#include <DD4hep/IDDescriptor.h>
+#include <DD4hep/Readout.h>
+#include <DD4hep/config.h>
+#include <DDSegmentation/BitFieldCoder.h>
+#include <Evaluator/DD4hepUnits.h>
+#include <algorithms/service.h>
+#include <fmt/core.h>
+#include <podio/RelationRange.h>
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <gsl/pointers>
+#include <limits>
+#include <stdexcept>
+#include <string>
+#include <map>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "algorithms/calorimetry/SimCalorimeterHitProcessorConfig.h"
+#include "services/evaluator/EvaluatorSvc.h"
+
+using namespace dd4hep;
+
+namespace eicrecon {
+
+void SimCalorimeterHitProcessor::init() {
+
+  // readout checks
+  if (m_cfg.readout.empty()) {
+    error("readoutClass is not provided, it is needed to know the fields in readout ids");
+    throw std::runtime_error("readoutClass is not provided");
+  }
+
+  // get decoders
+  try {
+    m_id_spec = m_geo.detector()->readout(m_cfg.readout).idSpec();
+  } catch (...) {
+    debug("Failed to load ID decoder for {}", m_cfg.readout);
+    throw std::runtime_error(fmt::format("Failed to load ID decoder for {}", m_cfg.readout));
+  }
+
+  uint64_t id_inverse_mask = 0;
+  // get id_mask for adding up hits that have the same z-segmentation
+  if (!m_cfg.mergeField.empty()) {
+    is_merge = true;
+    id_inverse_mask |= m_id_spec.field(m_cfg.mergeField)->mask();
+    debug("ID mask in {:s}: {:#064b}", m_cfg.readout, m_id_mask);
+  }
+  m_id_mask = ~id_inverse_mask;
+
+  // get attenuationMode for attenuating hits
+  if (!m_cfg.attenuationField.empty()) {
+    is_attenuation    = true;
+    m_attenuationMode = m_geo.detector()->constant<double>(m_cfg.attenuationField) / dd4hep::mm;
+  }
+}
+
+void SimCalorimeterHitProcessor::process(const SimCalorimeterHitProcessor::Input& input,
+                                         const SimCalorimeterHitProcessor::Output& output) const {
+
+  const auto [in_hits]              = input;
+  auto [out_hits, out_hit_contribs] = output;
+
+  // map for regrouping
+  std::map<edm4hep::MCParticle, std::vector<edm4hep::SimCalorimeterHit>> mapMCParToSimCalHit;
+
+  // regroup the sim hits by mc particle
+  for (const auto& ih : *in_hits) {
+    for (const auto& contrib : ih.getContributions()) {
```

Comment:
This grows potentially large. Can you provide some comment on how large you expect this to grow in worst case conditions? Does it affect memory limits on our running of reconstruction?